templates/commands/render.go

```diff
@@ -141,25 +161,155 @@ func (r *Render) parseFlags(args []string) error {
 	return nil
 }
 
+// We can't use os.DirFS or StatFS because they lack some methods we need. So we
+// implement our own filesystem interface.
+type realFS struct{}
+
+func (r *realFS) Stat(name string) (fs.FileInfo, error) {
+	return os.Stat(name) //nolint:wrapcheck
+}
+
+func (r *realFS) Open(name string) (fs.File, error) {
+	return os.Open(name) //nolint:wrapcheck
+}
+
+func (r *realFS) RemoveAll(name string) error {
+	return os.RemoveAll(name) //nolint:wrapcheck
+}
+
 func (r *Render) Run(ctx context.Context, args []string) error {
 	if err := r.parseFlags(args); err != nil {
 		return err
 	}
 
-	useFS := r.testFS // allow filesystem interaction to be faked for testing
-	if useFS == nil {
-		useFS = os.DirFS("/").(fs.StatFS) //nolint:forcetypeassert // safe per docs: https://pkg.go.dev/os#DirFS
+	fSys := r.testFS // allow filesystem interaction to be faked for testing
+	if fSys == nil {
+		fSys = &realFS{}
 	}
 
-	if err := destOK(useFS, r.flagDest); err != nil {
+	if err := destOK(fSys, r.flagDest); err != nil {
 		return err
 	}
 
-	return fmt.Errorf("not implemented")
+	gg := r.testGetter
+	if gg == nil {
+		gg = &getter.Client{
+			Getters:       getter.Getters,
+			Decompressors: getter.Decompressors,
+		}
+	}
+
+	if r.logger == nil {
+		r.logger = logging.NewFromEnv("ABC_")
+	}
```

Comment:
I don't think you need to make the logger a field. Once you have a logger, you can put it in context with https://github.com/abcxyz/pkg/blob/main/logging/logger.go#L86. During test, when you call `readRun`, you can put a test logger (https://github.com/abcxyz/pkg/blob/main/logging/logger.go#L109) in the context instead.

Reply:
I think this was my fault. I misled Dave with my comment "Can we embed it in the command or context instead?".

Reply:
Added to the context.